theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
I don't understand why creating new In() and Out() after having cloned the
fgraph to be difficult. I can just call copy.copy() on it and swap the
variable.

On Wed, May 27, 2015 at 12:16 PM, Chienli Ma(马千里) <notifications@github.com>
wrote:

> In theano/compile/function_module.py
> <https://github.com/Theano/Theano/pull/2969#discussion_r31151295>:
>
> > +        with original function. User can choose whether to separate storage by
> > +        changing the share_memory arguments
> > +        ---------------------
> > +        Params:
> > +            share_memory -- { boolean } Default is False. When True, two function
> > +            share intermediate storages(storages except input and output storages)
> > +        ---------------------
> > +        Returns:
> > +            func -- Copied theano.Function
> > +        """
> > +
> > +        if not share_memory:
> > +            return self.__copy__()
> > +        else:
> > +            # copy SymbolocKits
> > +            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
>
> Good idea. But there's another problem that variables in In() instances be
> the same as thoes in fgraph. It will be a little hard to rewarp In()
> instances or modify them with only an fgraph. So I decided to copy In()s
> first and then use variables inside to copy fgraph. I will try to find if
> there exist better ways.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2969/files#r31151295>.
>